valhalla/midgard/gridded_data.h

```diff
@@ -100,14 +116,339 @@ template <class coord_t> class GriddedData : public Tiles<coord_t> {
    *
    * @return contour line geometries with the larger intervals first (for rendering purposes)
    */
-  contours_t GenerateContours(const std::vector<float>& contour_intervals,
+  contours_t GenerateContours(std::vector<contour_interval_t>& intervals,
                               const bool rings_only = false,
                               const float denoise = 1.f,
-                              const float generalize = 200.f) const;
+                              const float generalize = 200.f) const {
+    // sort the contours first on the metric index then on the values with the bigger contours first
+    std::sort(intervals.begin(), intervals.end(), std::greater<>());
+
+    // Values at tile corners and center (0 element is center)
+    int sh[5];
+    typename PointLL::first_type s[5]; // Values at the tile corners and center
+    PointLL tile_corners[5];           // PointLL at tile corners and center
+    int m1, m2, m3;                    // Indices into the tile corners
+    PointLL pt1, pt2;                  // The intersection points in the tile
+    int tile_inc[4] = {0, 1, this->ncolumns_ + 1, this->ncolumns_};
+
+    // Find the intersection along a tile edge
+    auto intersect = [&tile_corners, &s](int p1, int p2) {
+      auto ds = s[p2] - s[p1];
+      return PointLL((s[p2] * tile_corners[p1].first - s[p1] * tile_corners[p2].first) / ds,
+                     (s[p2] * tile_corners[p1].second - s[p1] * tile_corners[p2].second) / ds);
+    };
+
+    // In the tight loop below, we need to decide where a contour intersects the triangles that make
+    // up the given tile. this works out to a number of discrete cases which we lookup using the table
+    // below. based on the case we perform the appropriate intersection. to avoid branching we store
+    // the intersection operation for each case in an array and perform the correct one by calling the
+    // function stored in the array
+    int case_table[3][3][3] = {
+        {{0, 0, 8}, {0, 2, 5}, {7, 6, 9}},
+        {{0, 3, 4}, {1, 3, 1}, {4, 3, 0}},
+        {{9, 6, 7}, {5, 2, 0}, {8, 0, 0}},
+    };
+    std::array<std::function<void()>, 10> cases{
```

Comment:
this was the other optimization. previously the below operations in the cases array were each an element of a switch case. now the call to them is a little strange looking below but we avoid all those branches in the tight loop